Create user preferences extensions and a page to view and update them

nemesis09 opened the pull request:
**Epic:**
https://issues.redhat.com/browse/ODC-5227

**Story:**
https://issues.redhat.com/browse/ODC-5939

**Analysis/Root cause:**
As a user, I need the ability to set/edit my preferences for the OCP Console from a designated place in the console.

**Solution/Description:**
This PR provides a dynamic extension to contribute user preferences. which are handled in a central UI component.
Three user preferences has been contributed as a part of this PR-

1. **Preferred Perspective:** used to set a preference for the perspective to be used as default across the console for the user. Prior to this we used last visited perspective.
2. **Preferred Namespace:** used to set a preference for the namespace to be used as default across the console. Prior to this we used last namespace.
3. **Preferred Topology View:** used to set a preference for the topology view. Prior to this, we used last used View type.

**Screens:**
![Screenshot from 2021-08-18 21-47-50](https://user-images.githubusercontent.com/38663217/129934733-c8a20a49-d93b-4fcd-9f94-2dd8837bbed8.png)
![Screenshot from 2021-08-18 21-50-26](https://user-images.githubusercontent.com/38663217/129934902-971aaab8-cdbd-4c5f-b9a7-986d2182648a.png)

**Test Coverage:**
![Screenshot from 2021-07-07 20-00-42](https://user-images.githubusercontent.com/38663217/124778786-dcd81280-df5e-11eb-9e3a-9aefecf6bddb.png)
![Screenshot from 2021-07-07 20-01-10](https://user-images.githubusercontent.com/38663217/124778792-de093f80-df5e-11eb-8b37-9a39fdb5cd57.png)
![Screenshot from 2021-07-07 20-01-29](https://user-images.githubusercontent.com/38663217/124778794-df3a6c80-df5e-11eb-88c1-67b26e841880.png)
![Screenshot from 2021-07-07 20-05-31](https://user-images.githubusercontent.com/38663217/124778797-dfd30300-df5e-11eb-98d0-2537ebac3825.png)

**Browser conformation:**
- [x] Firefox
- [x] Chrome
- [ ] Safari
- [ ] Edge

jerolimov commented:
/assign

jerolimov commented:
/assign

nemesis09 commented:
putting the PR on hold since there is a dependency on the PR https://github.com/openshift/console/pull/9279 for the component for namespace.
There also seems to be a change in the bookmarks and favorite user settings, which might affect the implementation of the preferred namespace as default.
This PR already has a similar namespace component since work on it was done before, so it can now be used for testing until we have the component from https://github.com/openshift/console/pull/9279.
cc @jerolimov 
/hold

nemesis09 commented:
removed user preference for namespace as it will be contributed in a follow up PR.
This PR includes-
-  the extension for user preferences
-  the user preference contributions for `Perspective` and `Topology View`

removing hold.
/hold cancel

cc @jerolimov

jerolimov commented on frontend/packages/console-app/console-extensions.json:
```diff
@@ -65,5 +65,51 @@
+          { "value": "dev", "label": "%console-app~Developer%" }
+        ]
+      },
+      "insertBefore": "console.preferredPerspective"
```
This is the `id` of the current element. Please drop this `insertBefore`

jerolimov commented on frontend/packages/console-app/console-extensions.json:
```diff
@@ -65,5 +65,51 @@
+    "type": "console.user-settings/group",
+    "properties": {
+      "id": "language",
+      "label": "%console-app~Language&region%",
```
Please add some spaces here:

```suggestion
      "label": "%console-app~Language & region%",
```

jerolimov commented on frontend/packages/console-app/console-extensions.json:
```diff
@@ -65,5 +65,51 @@
+    "properties": {
+      "id": "console.preferredPerspective",
+      "label": "%console-app~Perspective%",
+      "groupId": "general",
```
Can you reorder this and move (all) groups above the `preferredPerspective`? It's easier to understand if the refered groups comes before the reference.

jerolimov commented on frontend/packages/console-app/locales/en/console-app.json:
```diff
@@ -1,5 +1,13 @@
+  "Administrator": "Administrator",
+  "Developer": "Developer",
+  "General": "General",
+  "Language&region": "Language&region",
```
```suggestion
  "Language & region": "Language & region",
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.tsx:
```diff
@@ -0,0 +1,137 @@
+            title={<TabTitleText>{label}</TabTitleText>}
+            tabContentId={id}
+            tabContentRef={ref}
+            data-test={`userPreferenceTab ${id}`}
```
We don't want prefix these data. If there are more tabs on a screen we should add a data-test="user-preferences tabs" to the `Tabs` element.

```suggestion
            data-test={`tab ${id}`}
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.tsx:
```diff
@@ -0,0 +1,137 @@
+            id={id}
+            ref={ref}
+            hidden={id !== activeTabId}
+            data-test={`userPreferenceTabContent ${id}`}
```
We don't want prefix these data. If there are more tabs on a screen we should add a data-test="user-preferences tabs" to the `Tabs` element.

```suggestion
            data-test={`tab-content ${id}`}
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.tsx:
```diff
@@ -0,0 +1,137 @@
+                onSelect={handleTabClick}
+                isVertical
+                variant="light300"
+                data-test="userPreferenceTabs container"
```
```suggestion
                data-test="user-preferences tabs"
```

jerolimov reviewed (changes requested):
First feedback

jerolimov commented:
@nemesis09 Almost great implementation. Found some small issues:

1. When clicking the tab it sets the URL path, but when I click a second time it pushs another URL path which results in a 404:

![broken-tab-click](https://user-images.githubusercontent.com/139310/126462149-3eac1764-8d33-4f48-8bc0-3a58391539e1.gif)

2. When reducing the width or height there is no padding on the right side of the text fields or below the last text field. I expect that the complete content area should have a padding.

![missing-right-padding](https://user-images.githubusercontent.com/139310/126462381-d77ce935-08a2-4732-b49d-230e241e8fb4.gif)

![missing-bottom-padding](https://user-images.githubusercontent.com/139310/126462563-65559107-636f-456f-a174-0c3d104f34e7.gif)

jerolimov commented on frontend/packages/console-app/locales/en/console-app.json:
```diff
@@ -1,5 +1,13 @@
   "VolumeSnapshotContents": "VolumeSnapshotContents",
+  "Perspective": "Perspective",
+  "If a perspective is not selected, the console defaults to the last viewed.": "If a perspective is not selected, the console defaults to the last viewed.",
+  "Last Viewed": "Last Viewed",
```
```suggestion
  "Last viewed": "Last viewed",
```

jerolimov commented on frontend/packages/dev-console/console-extensions.json:
```diff
@@ -435,5 +435,26 @@
+        "options": [
+          {
+            "value": "latest",
+            "label": "%devconsole~Last Viewed%"
```
```suggestion
            "label": "%devconsole~Last viewed%"
```

jerolimov reviewed (changes requested):
Last Viewed => Last viewed

jerolimov commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,67 @@
+import * as React from 'react';
+import { Extension, ExtensionDeclaration, CodeRef } from '../types';
+
+export enum UserPreferencesInputType {
+  dropdown = 'dropdown',
+  checkbox = 'checkbox',
+}
+
+export type DropdownOption = {
+  type: UserPreferencesInputType.dropdown;
+  options: {
+    value: string;
+    label: string;
+  }[];
+};
+export type CheckboxOption = {
+  type: UserPreferencesInputType.checkbox;
+  trueValue: string;
+  falseValue: string;
+};
+export type UserSettingInputOptions = DropdownOption | CheckboxOption;
```
We should make a decision to call all these classes (and files, fields, ..) `UserPreferences...` or `UserSettings...`, and try to avoid mixing both. I'm fine with UserPreferences which is imho a more common term and navigation entry and page title calls this "User preferences".

Imho we should keep just `useUserSettings` and create a new issue to rename all useUserSettings hooks in another PR (to keep the focus here on the new components and still cleanup these later).

@christianvogt @invincibleJai fine with that or do you have any preference (hihi) ?

jerolimov replied:
Ah my idea was to call the individual preference field, because Aritra call the field options "inputOptions".

We have three levels with really similar name where it would be good to have as clean as possible names:

1. Groups
2. Items (or fields)
3. Item input definitions (or option) for each item (or field or option)

jerolimov commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,67 @@
+};
+export type UserSettingInputOptions = DropdownOption | CheckboxOption;
+
+export type UserSettings = ExtensionDeclaration<
+  'console.user-settings',
```
I would suggest to don't call this extension just "UserSettings" as this is a really generic term. Based on the decision if we want call all these types UserPreferences... or UserSettings... I would suggest something `UserPreferenceOption` or `UserPreferenceField`. Wdyt?

```suggestion
export type UserPreferenceField = ExtensionDeclaration<
  'console.user-settings/field',
```

jerolimov replied:
@christianvogt Any preference (hihi) here as well? :)

jerolimov replied:
It is *item now. Item or field is both for for me.

jerolimov replied:
In our call today we decided this extensions should be an item: `UserPreferenceItem`.

The field is the part of an item which describes just the input field (just the checkbox or the dropdown).

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.tsx:
```diff
@@ -0,0 +1,137 @@
+  return (
+    <>
+      <Helmet>
+        <title>{t('console-app~User Settings')}</title>
+      </Helmet>
+      <PageHeading detail title={t('console-app~User Settings')} />
```
The design document says the title should be "User preferences". There is also an subtitle below the page title in the design.

From design doc:

![image](https://user-images.githubusercontent.com/139310/126467817-8fbfb4f1-d3ea-450d-b81c-eb6ff6f3005e.png)

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.tsx:
```diff
@@ -0,0 +1,137 @@
+          <Tab
+            key={id}
+            eventKey={id}
+            title={<TabTitleText>{label}</TabTitleText>}
```
Please add a href here so that the user can ctrl-click another option.

```suggestion
            title={<TabTitleText>{label}</TabTitleText>}
            href={`/user-preferences/${id}`}
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.tsx:
```diff
@@ -0,0 +1,137 @@
+    );
+    return [tabs, tabContents];
+  };
+  const handleTabClick = (event, eventKey) => {
+    setActiveTabId(eventKey);
+    history.replace(`${url}/${eventKey}`);
+  };
```
```suggestion
  const handleTabClick = (event: React.MouseEvent<HTMLElement>, eventKey: string) => {
    if (isModifiedEvent(event)) {
      return;
    }
    setActiveTabId(eventKey);
    history.replace(`${url}/${eventKey}`);
  };
```

jerolimov reviewed (changes requested):
Please add a href to the `Tab`s so that the user can ctrl-click the tabs to open then in a new browser tab.

christianvogt commented on frontend/packages/console-app/src/components/user-preferences/utils/getTabContents.tsx:
```diff
@@ -0,0 +1,50 @@
+  if (inputOption.type === UserPreferencesInputType.checkbox) {
+    const { trueValue, falseValue } = inputOption;
+    return (
+      <FormGroup key={id} fieldId={id} label={label} data-test={id}>
```
Missing `helperText={description}` ?

nemesis09 replied:
yes. The design doc does have a use case for a Checkbox with `helperText`.
added `helperText`

christianvogt commented on frontend/packages/console-app/src/components/user-preferences/utils/getTabContents.tsx:
```diff
@@ -0,0 +1,50 @@
+    );
+  }
+
+  if (inputOption.type === UserPreferencesInputType.dropdown) {
+    return (
+      <FormGroup key={id} fieldId={id} label={label} helperText={description} data-test={id}>
+        <UserPreferenceInputDropdown
+          id={id}
+          label={label}
+          userSettingKey={id}
+          dropdownOptions={inputOption.options}
+        />
+      </FormGroup>
+    );
+  }
```
You should be able to create a map of type => component and then if present, simply render it without having to a big `if` condition per type.

nemesis09 commented:
> @nemesis09 Almost great implementation. Found some small issues:
> 
>     1. When clicking the tab it sets the URL path, but when I click a second time it pushs another URL path which results in a 404:
> 
> 
> ![broken-tab-click](https://user-images.githubusercontent.com/139310/126462149-3eac1764-8d33-4f48-8bc0-3a58391539e1.gif)
> 
>     1. When reducing the width or height there is no padding on the right side of the text fields or below the last text field. I expect that the complete content area should have a padding.
> 
> 
> ![missing-right-padding](https://user-images.githubusercontent.com/139310/126462381-d77ce935-08a2-4732-b49d-230e241e8fb4.gif)
> 
> ![missing-bottom-padding](https://user-images.githubusercontent.com/139310/126462563-65559107-636f-456f-a174-0c3d104f34e7.gif)
> 
>     1. If "console.preferredPerspective" is not defined the "Last viewed" option should be selected anyway. Because that is what the default option is. Maybe by adding a new attribute to define the default value? Like
> 
> 
> ```diff
>       "inputOption": {
>         "type": "dropdown",
> +       "defaultValue": "latest",
>         "options": [
>           {
>             "value": "latest",
>             "label": "%console-app~Last Viewed%"
>           },
>           { "value": "admin", "label": "%console-app~Administrator%" },
>           { "value": "dev", "label": "%console-app~Developer%" }
>         ]
>       },
> ```

1. Fixed issue with url being pushed with duplicate param.
2. Fixed paddings
3. Added a defaultValue param to the API

jerolimov commented on frontend/packages/console-app/console-extensions.json:
```diff
@@ -65,5 +65,51 @@
+            "label": "%console-app~Last Viewed%"
+          },
+          { "value": "admin", "label": "%console-app~Administrator%" },
+          { "value": "dev", "label": "%console-app~Developer%" }
```
Will there any ACM perspective in 4.9?

christianvogt replied:
Let's address this in the 4.9 timeframe. But it doesn't need to be in this pr.

jerolimov commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,67 @@
+    /** The input field options used to render the values to set the user setting. */
+    inputOption?: UserSettingInputOptions;
+    /** Custom component to set the user setting. */
+    customComponent?: CodeRef<React.ReactElement>;
```
I think that's fine. Custom components can use it or ignore it.

As discussed with Aritra, I would recommend a `UserPreferenceFormGroup` component which renders the PF `FormGroup` and call `useUserSettings`. And then the predefined field components (as well as the custom component) will have props similar to this:

```tsx
type UserPreferencesCustomFieldProps<Value> = {
  field: UserPreferenceField;
  value: Value;
  onChange: (newValue: Value) => void;
}
```

christianvogt commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.scss:
```diff
@@ -0,0 +1,16 @@
+.ocs-user-preferences-content {
```
we are using `co` in `console-app`

christianvogt commented on frontend/packages/console-app/src/components/user-preferences/UserPreferences.tsx:
```diff
@@ -0,0 +1,157 @@
+      >
+        {userPreferencesItemResolved ? (
+          <div className="ocs-user-preferences-content">
+            <div id="tabs-section" className="ocs-user-preferences-content__tabs">
```
Can we remove this `id` and the one below?

nemesis09 replied:
removed

nemesis09 commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,67 @@
+    /** The input field options used to render the values to set the user setting. */
+    inputOption?: UserSettingInputOptions;
+    /** Custom component to set the user setting. */
+    customComponent?: CodeRef<React.ReactElement>;
```
updated the type to `field`

nemesis09 commented:
putting on hold since it depends on https://github.com/openshift/console/pull/9070 for resolving nested `CodeRefs`
cc @jerolimov @christianvogt 
/hold

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/perspective/usePreferredPerspective.ts:
```diff
@@ -0,0 +1,10 @@
+import { useUserSettings } from '@console/shared';
+
+const PREFERRED_PERSPECTIVE_USER_SETTING_KEY: string = 'console.preferredPerspective';
```
You don't need add unambiguous types when defining a string const.

This is maybe a good place for a const for "latest". Maybe we can change then the value to `%LATEST%` similar to ALL_NAMESPACES_KEY to ensure that we don't have a problem with a project which is called "latest". :)

```suggestion
const PREFERRED_PERSPECTIVE_USER_SETTING_KEY = 'console.preferredPerspective';
const PREFERRED_PERSPECTIVE_USER_SETTING_LATEST = '%LATEST%';
```

jerolimov commented on frontend/packages/console-app/console-extensions.json:
```diff
@@ -76,5 +76,52 @@
+      "field": {
+        "type": "dropdown",
+        "userSettingsKey": "console.preferredPerspective",
+        "defaultValue": "latest",
```
Similar to ALL_NAMESPACES_KEY to ensure that we don't have an issue with a perspective (and use a similar way when doing this for namespaces later). :)

```suggestion
        "defaultValue": "%LATEST%",
```

christianvogt replied:
Do we even need a value here? If the value is undefined, can we treat it as latest?

jerolimov commented on frontend/packages/console-app/console-extensions.json:
```diff
@@ -76,5 +76,52 @@
+        "defaultValue": "latest",
+        "options": [
+          {
+            "value": "latest",
```
Similar to ALL_NAMESPACES_KEY to ensure that we don't have an issue with a perspective (and use a similar way when doing this for namespaces later). :)

```suggestion
            "value": "%LATEST%",
```

jerolimov commented on frontend/packages/console-app/src/components/detect-perspective/useValuesForPerspectiveContext.ts:
```diff
@@ -13,12 +14,19 @@ export const useValuesForPerspectiveContext = (): [
+  const loaded = lastPerspectiveLoaded && preferredPerspectiveLoaded;
+
+  const perspective =
+    loaded && preferredPerspective && preferredPerspective !== 'latest'
```
Const see below (should be replace all 'latest')

```suggestion
    loaded && preferredPerspective && preferredPerspective !== PREFERRED_PERSPECTIVE_USER_SETTING_LATEST
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default as UserPreference } from './UserPreferencePage';
```
I would recommend to not rename components. Make it easier to find the right component / file.

```suggestion
export { default as UserPreferencePage } from './UserPreferencePage';
```

nemesis09 replied:
missed that while renaming. Thanks!

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferenceCustomField.tsx:
```diff
@@ -0,0 +1,11 @@
+
+type UserPreferenceCustomFieldProps = UserPreferenceFieldProps<CustomFieldType>;
+
+const UserPreferenceCustomField: React.FC<UserPreferenceCustomFieldProps> = ({
+  component: Component,
+  props,
+}) => (Component ? <Component {...props} /> : null);
```
Only if you want make it more clear that props is a specific attribute here:

```suggestion
const UserPreferenceCustomField: React.FC<UserPreferenceCustomFieldProps> = ({
  component: Component,
  props: customComponentProps,
}) => (Component ? <Component {...customComponentProps} /> : null);
```

nemesis09 replied:
If we are renaming it to `customComponentProps`, I guess we could also update from `Component` to `CustomComponent`.
WDYT?

jerolimov replied:
I like that idea.

jerolimov commented on frontend/packages/topology/user-preferences/usePreferredTopologyView.ts:
```diff
@@ -0,0 +1,10 @@
+import { useUserSettings } from '@console/shared';
+
+const PREFERRED_TOPOLOGY_VIEW_USER_SETTING_KEY: string = 'topology.preferredView';
```
```suggestion
const PREFERRED_TOPOLOGY_VIEW_USER_SETTING_KEY = 'topology.preferredView';
```

jerolimov commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,81 @@
+  type: UserPreferenceFieldType.checkbox;
+  userSettingsKey: string;
+  label: string;
+  trueValue: string;
+  falseValue: string;
```
I don't think this must be strings. It's maybe any(thing) which could be defined in JSON. But at least we should start with `string | number | boolean`. Wdyt?

```suggestion
  trueValue: string | number | boolean;
  falseValue: string | number | boolean;
```

nemesis09 replied:
yes. we can have non string values for true and false values. Updated types.

jerolimov commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,81 @@
+export type UserPreferenceCustomField = {
+  type: UserPreferenceFieldType.custom;
+  component: CodeRef<React.ComponentType>;
+  props?: { [key: string]: string };
```
I don't like any. But I think in the JSON the user could define more then just string: string.

```suggestion
  props?: { [key: string]: any };
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferenceCheckboxField.tsx:
```diff
@@ -0,0 +1,45 @@
+
+  const loaded: boolean = currentUserPreferenceValueLoaded;
+
+  // utils and callbacks
+  const getCheckboxValue = (checked: boolean): string => (checked ? trueValue : falseValue);
+
+  const onChange = (checked) => {
+    const checkedValue: string = getCheckboxValue(checked);
+    checkedValue !== currentUserPreferenceValue && setCurrentUserPreferenceValue(checkedValue);
+  };
```
Don't force that `trueValue` and `falseValue` are strings.

```suggestion
  // utils and callbacks
  const getCheckboxValue = (checked: boolean) => (checked ? trueValue : falseValue);

  const onChange = (checked) => {
    const checkedValue = getCheckboxValue(checked);
    checkedValue !== currentUserPreferenceValue && setCurrentUserPreferenceValue(checkedValue);
  };
```

nemesis09 replied:
updated types

jerolimov reviewed (commented):
Small improvement ideas. In general it looks really good :+1:

nemesis09 commented on frontend/packages/console-app/console-extensions.json:
```diff
@@ -76,5 +76,52 @@
+    "properties": {
+      "id": "general",
+      "label": "%console-app~General%",
+      "insertBefore": "language"
```
yes. we don't need these attributes when they are in order.
Removed.

jerolimov commented on frontend/public/components/utils/list-dropdown.jsx:
```diff
@@ -204,7 +204,7 @@ ListDropdown.propTypes = {
 };
 
-const useProjectOrNamespaceModel = () => {
+export const useProjectOrNamespaceModel = () => {
```
Maybe a work in process change? At the moment this change is not needed.

```suggestion
const useProjectOrNamespaceModel = () => {
```

nemesis09 replied:
This is needed to decide the use of `Namespace` or `Project` in string literals in the `NamespaceDropdown` component

jerolimov commented on frontend/packages/console-app/src/components/detect-perspective/__tests__/useValuesForPerspectiveContext.spec.ts:
```diff
@@ -0,0 +1,81 @@
+  it('should return undefined for PerspectiveType if the perspective from user settings does not match any of the perspectives received from extensions', () => {
+    useExtensionsMock.mockReturnValue(mockPerspectiveExtensions);
+    useUserSettingsCompatibilityMock.mockReturnValue(['acm', () => {}, true]);
+    usePreferredPerspectiveMock.mockReturnValue(['latest', true]);
```
Did you want return latest here? It returns undefined anyway because acm perspective is not defined, right?

```suggestion
    usePreferredPerspectiveMock.mockReturnValue([PREFERRED_PERSPECTIVE_USER_SETTING_VALUE_LATEST, true]);
```

nemesis09 replied:
missed this. Thanks!

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferenceCheckboxField.tsx:
```diff
@@ -0,0 +1,49 @@
+
+  const loaded: boolean = currentUserPreferenceValueLoaded;
+
+  // utils and callbacks
+  const getCheckboxValue = (checked: boolean): UserPreferenceValueType =>
+    checked ? trueValue : falseValue;
+
+  const onChange = (checked) => {
+    const checkedValue: UserPreferenceValueType = getCheckboxValue(checked);
+    checkedValue !== currentUserPreferenceValue && setCurrentUserPreferenceValue(checkedValue);
+  };
```
Could be simplified into one function. Value is more needed for the onChange argument where the compiler could not know it then for the const value .... where the type is known based on the type of `? trueValue : falseValue`.

```suggestion
  const onChange = (checked: boolean) => {
    const value = checked ? trueValue : falseValue;
    if (value !== currentUserPreferenceValue) {
      setCurrentUserPreferenceValue(value);
    }
  };
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferencePage.scss:
```diff
@@ -0,0 +1,16 @@
+  flex-direction: row;
+  background-color: var(--pf-global--BackgroundColor--100);
+  height: 100%;
+  padding: 0 var(--pf-global--spacer--lg) var(--pf-global--spacer--lg) 0;
```
Instead of adding magic padding we should ensure that the tabs columns have the right width:

```suggestion
  &__tabs {
    flex: 0 0 220px; // similar to catalog
  }
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferencePage.scss:
```diff
@@ -0,0 +1,16 @@
+  height: 100%;
+  padding: 0 var(--pf-global--spacer--lg) var(--pf-global--spacer--lg) 0;
+  &__tab-content {
+    margin: var(--pf-global--spacer--lg) 0 0 var(--pf-global--spacer--3xl);
```
The content width is already depending on the description length (help text below the select)

I believe we should give all content elements the full width instead. If you or UX doesn't like that we can skip `flex: 1` here. But with the simplified padding above the margin could be simplified here as well:

```suggestion
    flex: 1;
    margin: var(--pf-global--spacer--lg);
```

nemesis09 replied:
As discussed, we could wrap the description instead, similar to what is done in the import forms.

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferencePage.scss:
```diff
@@ -0,0 +1,16 @@
+  padding: 0 var(--pf-global--spacer--lg) var(--pf-global--spacer--lg) 0;
+  &__tab-content {
+    margin: var(--pf-global--spacer--lg) 0 0 var(--pf-global--spacer--3xl);
+    &__item {
+        margin-top: var(--pf-global--spacer--lg);
+    }
```
Not used, remove `__item`

```suggestion
```

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferencePage.scss:
```diff
@@ -0,0 +1,16 @@
+        margin-top: var(--pf-global--spacer--lg);
+    }
+  }
+  .pf-c-tabs__scroll-button {
+    display: none;
+  }
```
Okay got why you want drop these buttons. I believe its an error that the buttons are rendered when `isVertical` prop is true, or?

Can you please add a comment and if you want also open an issue at https://github.com/patternfly/patternfly-react

jerolimov commented on frontend/packages/console-app/src/components/user-preferences/UserPreferencePage.scss:
```diff
@@ -0,0 +1,16 @@
+  display: flex;
+  flex-direction: row;
+  background-color: var(--pf-global--BackgroundColor--100);
+  height: 100%;
```
Don't force the height, I believe you want ask for a minimum height. ;)

```suggestion
  min-height: 100%;
```

nemesis09 commented on frontend/packages/console-app/src/components/user-preferences/UserPreferencePage.tsx:
```diff
@@ -0,0 +1,154 @@
+
+  // utils and callbacks
+  const getTabs = (): [React.ReactElement<TabProps>[], React.ReactElement<TabContentProps>[]] => {
+    const populatedUserPreferenceGroups: UserPreferenceTabGroup[] = getUserPreferenceGroups(
+      sortedUserPreferenceGroups,
+      sortedUserPreferenceItems,
+    );
+    const [tabs, tabContents] = populatedUserPreferenceGroups.reduce(
+      (acc, currGroup) => {
+        const { id, label, items } = currGroup;
+        const ref = React.createRef<HTMLElement>();
+        acc[0].push(
+          <Tab
+            key={id}
+            eventKey={id}
+            title={<TabTitleText>{label}</TabTitleText>}
+            href={`${USER_PREFERENCES_BASE_URL}/${id}`}
+            tabContentId={id}
+            tabContentRef={ref}
+            data-test={`tab ${id}`}
+          />,
+        );
+        acc[1].push(
+          <TabContent
+            key={id}
+            eventKey={id}
+            id={id}
+            ref={ref}
+            hidden={id !== activeTabId}
+            data-test={`tab-content ${id}`}
+          >
+            <UserPreferenceForm items={items} />
+          </TabContent>,
+        );
+        return acc;
+      },
+      [[], []],
+    );
+    return [tabs, tabContents];
+  };
```
Interesting. Yes, we could do this in a `useMemo` instead of an util. It would be useful when we have many groups and items.

nemesis09 commented on frontend/packages/console-app/src/components/user-preferences/UserPreferenceForm.tsx:
```diff
@@ -0,0 +1,16 @@
+
+type UserPreferenceFormProps = { items: ResolvedUserPreferenceItem[] };
+
+const UserPreferenceForm: React.FC<UserPreferenceFormProps> = ({ items }) =>
+  items?.length > 0 ? (
+    <Form>
+      {items.map((item) => (
+        <UserPreferenceField key={item.id} item={item} />
+      ))}
+    </Form>
+  ) : null;
```
Thanks for the catch. updated to avoid default behavior.

jerolimov commented:
/hold cancel

jerolimov commented:
/test e2e-gcp-console

jerolimov commented:
/retest

jerolimov commented:
/retest

christianvogt commented on frontend/packages/topology/src/components/page/TopologyPage.tsx:
```diff
@@ -93,16 +110,16 @@ export const TopologyPage: React.FC<TopologyPageProps> = ({
         })
       ? TopologyViewType.graph
-      : isTopologyViewStateLoaded && (topologyViewState || defaultViewType);
+      : loaded && ((getTopologyViewState() as TopologyViewType) || defaultViewType);
```
Why the cast to `TopologyViewType` here? the function should return the appropriate type.

christianvogt commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,84 @@
+export type UserPreferenceDropdownField = {
+  type: UserPreferenceFieldType.dropdown;
+  userSettingsKey: string;
+  defaultValue?: string;
```
should this be optional?
What happens if there is no default value provided; will it render an empty selection?

nemesis09 replied:
IMHO, the default should be optional. If there is no default value provided, then we currently show no selection which means no user preference has been set, which in my opinion is a valid scenario. In such a case we render the placeholder which asks the user to make a selection.

christianvogt commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,84 @@
+
+export type UserPreferenceCheckboxFieldValue = string | number | boolean;
+
+export type UserPreferenceCheckboxField = {
```
Missing `defaultValue`?

nemesis09 replied:
added defaultValue

christianvogt commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,84 @@
+export type UserPreferenceCustomField = {
+  type: UserPreferenceFieldType.custom;
+  component: CodeRef<React.ComponentType>;
+  // TODO: update type from any to json compatible types
+  props?: { [key: string]: any };
```
What about `JSONSchema7` from `json-schema`?

nemesis09 replied:
If I'm not getting it wrong, I think you mean `JSONSchema7Type`, and i think it will be a great fit for the prop type.

jerolimov commented:
/retest

nemesis09 commented:
/retest

jerolimov commented:
Since the epic acceptance criteria of [ODC-5227](https://issues.redhat.com/browse/ODC-5227) have not changed and the epic has a doc-ack and px-ack, we approve this on behalf of doc and PX team.

/label doc-approved
/label px-approved

jerolimov commented:
Since the epic acceptance criteria of [ODC-5227](https://issues.redhat.com/browse/ODC-5227) have not changed and the epic has a doc-ack we approve this on behalf of doc team.

/label docs-approved

jerolimov reviewed (approved):
Well done, works as expected :+1:

jerolimov commented:
Tested it also on a cluster and works for me.

/label qe-approved

christianvogt commented on frontend/packages/console-app/src/components/user-preferences/perspective/usePreferredPerspective.ts:
```diff
@@ -0,0 +1,11 @@
+import { useUserSettings } from '@console/shared';
+
+export const PREFERRED_PERSPECTIVE_USER_SETTING_VALUE_LATEST = '#LATEST#';
+export const PREFERRED_PERSPECTIVE_USER_SETTING_KEY = 'console.preferredPerspective';
+
+export const usePreferredPerspective = (): [string, boolean] => {
+  const [preferredPerspective, , preferredPerspectiveLoaded] = useUserSettings<string>(
+    PREFERRED_PERSPECTIVE_USER_SETTING_KEY,
+  );
+  return [preferredPerspective, preferredPerspectiveLoaded];
+};
```
Same comments as the similarly implemented namespace hook.

christianvogt commented on frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts:
```diff
@@ -0,0 +1,85 @@
+export type UserPreferenceItem = ExtensionDeclaration<
+  'console.user-preference/item',
+  {
+    /** ID used to identify the user preference item. The value for ID is generally the key for the user preference in config map */
+    id: string;
```
Didn't we say that the preference key in the configmap for the pre-defined primitive fields would actually be a property named `key` on the `field` itself and not the `id`? 
This way we'd separate out the two concerns and only use the ID for ordering?
cc @jerolimov

jerolimov replied:
That's right. The comment is wrong. We have a `userSettingsKey` defined in the field attribute.

In `UserPreferenceDropdownField` and `UserPreferenceCheckboxField` which could maybe be unified, but for me that's not so important at the moment. But Christian is right we should change the comment here:

```suggestion
    /** ID used to identify the user preference item, referenced in insertAfter and insertBefore to define the item order. */
    id: string;
```

christianvogt commented on frontend/packages/console-app/src/components/detect-namespace/namespace.ts:
```diff
@@ -3,90 +3,89 @@ import * as React from 'react';
+  }, [resourcesLoaded, useProjects]);
+
+  React.useEffect(() => {
+    if (urlNamespace) {
+      getValueForNamespace(useProjects, urlNamespace)
```
Don't you also need to wait for `!flagPending(useProjects)` ?

nemesis09 replied:
yes. added it to `resourceLoaded`.
Thanks for catching this.

christianvogt commented on frontend/packages/console-app/src/components/detect-namespace/namespace.ts:
```diff
@@ -3,90 +3,89 @@ import * as React from 'react';
-  }, [favoriteLoaded, lastNamespaceLoaded, useProjects]);
+  }, [resourcesLoaded, useProjects]);
+
+  React.useEffect(() => {
+    if (urlNamespace) {
+      getValueForNamespace(useProjects, urlNamespace)
+        .then((ns) => {
+          setActiveNamespace(ns);
+          // sync with redux store
+          dispatch(setActiveNamespaceForStore(ns));
+        })
+        .catch((e) => {
+          // eslint-disable-next-line no-console
+          console.warn('Error fetching namespace', e);
+        });
+    }
+  }, [dispatch, urlNamespace, useProjects]);
```
Won't this effect race against the other effect?

jerolimov replied:
@christianvogt Did you think it is necessary to check if the `urlNamespace` exists? For me it would be okay to show an "Project not found" error in this case instead of magically redirect the user when the user opens a specific namespaced link..

If this is okay we could use urlNamespace as default value for the activeNamespace state and don't need a second useEffect.

christianvogt replied:
The current implementation does the redirect and therefore I'm reluctant to change the behavior at this time.

Can we combine both effects together like it was done before?

nemesis09 replied:
As far as I could understand we would still need a `useEffect`, since router links don't reload the console, so the state will not be reinitialized with the urlNamespace.
We can add a condition to avoid a race condition.
something like 
` React.useEffect(() => {
if (activeNamespace && urlNamespace && urlNamespace !== activeNamespace) {
// update state
}
`
WDYT?

nemesis09 commented on frontend/packages/console-app/src/components/detect-namespace/getValueForNamespace.ts:
```diff
@@ -0,0 +1,37 @@
+  lastNamespace?: string,
+): Promise<string> => {
+  if (urlNamespace) {
+    if (await checkNamespaceExists(urlNamespace, useProjects)) {
+      return urlNamespace;
+    }
```
I was thinking that it shouldn't throw since any error thrown is caught in `checkNamespaceExists`.
Please let me know if I am missing a scenario here.

christianvogt replied:
looks like `checkNamespaceExists` will never throw. So this is fine.

nemesis09 commented:
/retest

jerolimov commented:
/retest

nemesis09 commented:
/retest

nemesis09 commented:
/retest

nemesis09 commented:
/retest

nemesis09 commented:
/retest

nemesis09 commented:
/retest

nemesis09 commented:
/retest

nemesis09 commented:
/test e2e-gcp-console

nemesis09 commented:
/retest

nemesis09 commented:
/retest

nemesis09 commented:
/retest

rohitkrai03 commented:
/retest

jerolimov reviewed (approved):
/lgtm
